src/spikeinterface_pipelines/visualization/visualization.py

```diff
@@ -0,0 +1,237 @@
+
+from __future__ import annotations
+
+from pathlib import Path
+import numpy as np
+
+import matplotlib
+import matplotlib.pyplot as plt
+from matplotlib.colors import Normalize
+
+import spikeinterface.full as si
+import spikeinterface.widgets as sw
+
+import sortingview.views as svv
+
+import kachery_cloud as kcl
+from spikeinterface.widgets import sorting_summary
+
+from ..logger import logger
+
+from .params import VisualizationParams
+
+matplotlib.use("agg")
+
+
+def visualize(
+    recording: si.BaseRecording,
+    sorting_curated: si.BaseSorting | None = None,
+    waveform_extractor: si.WaveformExtractor | None = None,
+    visualization_params: VisualizationParams = VisualizationParams(),
+    scratch_folder: Path = Path("./scratch/"),
+    results_folder: Path = Path("./results/visualization/"),
+) -> dict | None:
+    """ 
+    Generate visualization of preprocessing, spikesorting and curation results.
+
+    Parameters
+    ----------
+    recording_processed: si.BaseRecording
+        The input processed recording
+    sorting_curated: si.BaseSorting | None
+        The input curated sorting. If None, only the recording visualization will be generated.
+    waveform_extractor: si.WaveformExtractor | None
+        The input waveform extractor from postprocessing. If None, only the recording visualization will be generated.
+    visualization_params: VisualizationParams
+        The visualization parameters
+    scratch_folder: Path
+        The scratch folder
+    results_folder: Path
+        The results folder
+
+    Returns
+    -------
+    visualization_output: dict
+        The visualization output dictionary    
+    """
+    logger.info("[Visualization] \tRunning Visualization stage")
+    visualization_output = {}
+    results_folder.mkdir(exist_ok=True, parents=True)
+
+    if kcl.get_client_info() is None:
+        logger.info(
+            "[Visualization] \tKachery client not found. Use `kachery-cloud-init` to initialize kachery client."
+        )
+        return
+    visualization_params_dict = visualization_params.model_dump()
+    recording_params = visualization_params_dict["recording"]
+
+    # Recording visualization
+    cmap = plt.get_cmap(recording_params["drift"]["cmap"])
+    norm = Normalize(
+        vmin=recording_params["drift"]["vmin"], vmax=recording_params["drift"]["vmax"], clip=True
+    )
+    decimation_factor = recording_params["drift"]["decimation_factor"]
+    alpha = recording_params["drift"]["alpha"]
+
+    # use spike locations
+    if waveform_extractor is not None and waveform_extractor.is_extension("spike_locations"):
```

Comment:
```suggestion
    if waveform_extractor is not None and waveform_extractor.has_extension("spike_locations"):
```
```suggestion
    if not waveform_extractor.has_extension("quality_metrics"):
```
with 0.100 this is the correct syntax